browser/brave_ads/tabs/html_content_body_handler.cc

```diff
@@ -0,0 +1,186 @@
+  if (!web_contents) {
+    return nullptr;
+  }
+
+  return web_contents->GetBrowserContext();
+}
+
+bool IsInterstitialOrErrorPageNavigation(
+    content::NavigationEntry* navigation_entry) {
+  CHECK(navigation_entry);
+  return navigation_entry->GetPageType() == content::PAGE_TYPE_ERROR;
+}
+
+}  // namespace
+
+HtmlContentBodyHandler::HtmlContentBodyHandler(
+    AdsService* const ads_service,
+    content::WebContents::Getter web_contents_getter)
+    : ads_service_(ads_service),
+      web_contents_getter_(std::move(web_contents_getter)) {
+  CHECK(ads_service_);
+}
+
+HtmlContentBodyHandler::~HtmlContentBodyHandler() = default;
+
+// static
+std::unique_ptr<HtmlContentBodyHandler> HtmlContentBodyHandler::MaybeCreate(
+    AdsService* const ads_service,
+    const content::WebContents::Getter& web_contents_getter) {
+  content::BrowserContext* const browser_context =
+      GetBrowserContext(web_contents_getter);
+  if (!browser_context || browser_context->IsOffTheRecord()) {
+    return {};
+  }
+
+  return base::WrapUnique(
+      new HtmlContentBodyHandler(ads_service, web_contents_getter));
+}
+
+bool HtmlContentBodyHandler::OnRequest(
+    network::ResourceRequest* const request) {
+  CHECK(request);
+
+  const bool is_main_frame =
+      request->resource_type ==
+      static_cast<int>(blink::mojom::ResourceType::kMainFrame);
+  if (!is_main_frame) {
+    // Don't notify content changes for subframes.
+    return false;
+  }
+
+  content::NavigationEntry* const navigation_entry =
+      GetPendingNavigationEntry();
+  if (!navigation_entry || navigation_entry->IsRestored() ||
+      IsInterstitialOrErrorPageNavigation(navigation_entry) ||
+      !ui::PageTransitionIsNewNavigation(
+          navigation_entry->GetTransitionType())) {
+    // Don't notify content changes if the tab was restored, was a previously
+    // committed navigation, or an error page was displayed.
+    return false;
+  }
+
+  return true;
+}
+
+bool HtmlContentBodyHandler::ShouldProcess(
+    const GURL& response_url,
+    network::mojom::URLResponseHead* const response_head,
+    bool* const defer) {
+  CHECK(response_head);
+  CHECK(defer);
+
+  std::string mime_type;
+  if (!response_head || !response_head->headers ||
+      !response_head->headers->GetMimeType(&mime_type) ||
+      !base::EqualsCaseInsensitiveASCII(mime_type, "text/html")) {
+    return false;
+  }
+
+  if (HttpResponseHasErrorCode(response_head->headers.get())) {
+    // Don't notify HTML content changes if an error page was displayed.
+    return false;
+  }
+
+  *defer = false;
+
+  return true;
+}
+
+void HtmlContentBodyHandler::OnComplete() {
+  MaybeNotifyTabHtmlContentDidChange();
+  html_ = std::string();
+}
+
+HtmlContentBodyHandler::Action HtmlContentBodyHandler::OnBodyUpdated(
+    const std::string& body,
+    const bool is_complete) {
+  if (!is_complete) {
+    return HtmlContentBodyHandler::Action::kContinue;
+  }
+
+  html_ = body;
```

Comment:
need to cache html content and send it in OnComplete, because navigation is not committed at this time, and we cannot get redirect chain.